rollout/trafficrouting/traefik/traefik.go

```diff
@@ -0,0 +1,164 @@
+}
+
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	reconciler := &Reconciler{
+		Rollout: cfg.Rollout,
+		Client:  cfg.Client,
+	}
+	return reconciler
+}
+
+func NewDynamicClient(di dynamic.Interface, namespace string) dynamic.ResourceInterface {
+	return di.Resource(GetMappingGVR()).Namespace(namespace)
+}
+
+func GetMappingGVR() schema.GroupVersionResource {
+	group := defaults.DefaultTraefikAPIGroup
+	parts := strings.Split(defaults.DefaultTraefikVersion, "/")
+	version := parts[len(parts)-1]
+	resourceName := apiGroupToResource[group]
+	return schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resourceName,
+	}
+}
+
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	return nil
+}
+
+func (r *Reconciler) SetWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) error {
+	ctx := context.TODO()
+	rollout := r.Rollout
+	traefikServiceName := rollout.Spec.Strategy.Canary.TrafficRouting.Traefik.TraefikServiceName
+	traefikService, err := r.Client.Get(ctx, traefikServiceName, metav1.GetOptions{})
+	if err != nil || traefikService == nil {
+		return err
+	}
+	canaryServiceName := rollout.Spec.Strategy.Canary.CanaryService
+	stableServiceName := rollout.Spec.Strategy.Canary.StableService
+	services, isFound, err := unstructured.NestedSlice(traefikService.Object, "spec", "weighted", "services")
+	if err != nil || !isFound {
+		return err
+	}
+	canaryService, err := getService(canaryServiceName, services)
+	if err != nil || canaryService == nil {
+		return err
+	}
+	err = unstructured.SetNestedField(canaryService, int64(desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	stableService, err := getService(stableServiceName, services)
+	if err != nil || stableService == nil {
+		return err
+	}
+	err = unstructured.SetNestedField(stableService, int64(100-desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	err = unstructured.SetNestedSlice(traefikService.Object, services, "spec", "weighted", "services")
+	if err != nil {
+		return err
+	}
+	_, err = r.Client.Update(ctx, traefikService, metav1.UpdateOptions{})
+	return err
+}
+
+func getService(serviceName string, services []interface{}) (map[string]interface{}, error) {
+	var selectedService map[string]interface{}
+	for _, service := range services {
+		typedService, ok := service.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("Failed type assertion setting weight for traefik service")
+		}
+		nameOfCurrentService, isFound, err := unstructured.NestedString(typedService, "name")
+		if err != nil || !isFound {
+			return nil, err
+		}
+		if nameOfCurrentService == serviceName {
+			selectedService = typedService
+			break
+		}
+	}
+	return selectedService, nil
+}
+
+func (r *Reconciler) VerifyWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) (*bool, error) {
```

Comment:
Probably there is no need to implement this method in that way. You're accessing the same traefic service object (which you just modified in a SetWeight method)

Right now it's implemented for ALB only. The reason whay we have this implementation for the ALB is delays in traffic shifting.
For the ALB once you modified the Ingress object it take some time to really shift the traffic by creating/modifying target group. So here we need to make validation against the target group objects to see the traffic was really updated.